lib/api/ldp/ldp-operation.js

```diff
@@ -0,0 +1,196 @@
+'use strict'
+
+const url = require('url')
+const glob = require('glob')
+// const HttpError = require('standard-http-error')
+
+const BY_METHOD = {
+  'head': LdpHeadOperation,
+  'get': LdpGetOperation,
+  'put': LdpPutOperation,
+  'post': LdpPostOperation,
+  'patch': LdpPatchOperation,
+  'delete': LdpDeleteOperation,
+  'copy': LdpCopyOperation,
+  'options': LdpOptionsOperation
+}
+
+class LdpOperation {
+  /**
+   * @param options {object}
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.headers {object}
+   *
+   * @param [options.authentication=null] {Authentication} Authn object,
+   *   contains WebID string, as well as any bearer credentials/tokens. Needed
+   *   for authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.bodyStream] {Stream} Request body stream, for parsing when
+   *   needed
+   */
+  constructor (options) {
+    this.authentication = options.authentication
+    this.target = options.target
+    this.headers = options.headers
+    this.bodyStream = options.bodyStream
+  }
+
+  /**
+   * General construction method, overridden where needed (such as in
+   * LdpGetOperation).
+   *
+   * @param options {object}
+   *
+   * @returns {LdpOperation}
+   */
+  static from (options) {
+    const Operation = this
+    return new Operation(options)
+  }
+
+
+  writeCommonHeaders ({res}) {
+    // set headers in common to all LDP responses
+  }
+}
+
+LdpOperation.BY_METHOD = BY_METHOD
+
+/**
+ * Checks to see if target exists
+ */
+class LdpHeadOperation extends LdpOperation {
+  writeHeaders ({res, result, permissions}) {
+    this.writeCommonHeaders({res})
+    // write HEAD-specific headers
+  }
+}
+
+/**
+ * Use cases to handle:
+ *  - "List Container" request, if target is an ldp container
+ *  - However, if Accept: header is HTML, check if index.html exists
+ *  - Need to also support similar use case if index.ttl exists
+ *  - "Glob pattern request" if glob magic chars are detected (ie `*`)
+ *  - Otherwise, plain Get request
+ *
+ *  @return {LdpGlobOperation|LdpListContainerOperation|LdpGetOperation}
+ */
+class LdpGetOperation extends LdpOperation {
+  static from (options) {
```

Comment:
@RubenVerborgh ok, so, I'm not *super* thrilled with this factory method, but it at least roughly fleshes out our semantics.

Reply:
Indeed, it's quite alright!